compiler/z/codegen/OMRTreeEvaluator.cpp

```diff
@@ -14117,6 +14117,168 @@ bool isSettingOp(uint8_t byteValue, TR::InstOpCode::Mnemonic SI_opcode)
+            // the first 6 bits of the operand as the shift amount. So no mask is required in that case
+            generateRIInstruction(cg, TR::InstOpCode::NILL, node, tmpReg, 0x1f);   
+            //TODO: There should be two of these
+            }
+         //create memory reference using tmpReg (shift amount), then shift valueReg by by shift amount
+         // TODO: Should another register be used here instead of tmpReg?
+         TR::MemoryReference * tempMR = generateS390MemoryReference(tmpReg, 0, cg);
+         generateRSInstruction(cg, TR::InstOpCode::getShiftRightLogicalSingleOpCode(), node, tmpReg, valueReg, tempMR); 
+         tempMR->stopUsingMemRefRegister(cg); 
+
+         if (node->getDataType() != TR::Int64)
+            {
+            // now and the value in tmpReg by 1 to get your relevant bit
+            generateRILInstruction(cg, TR::InstOpCode::NILF, node,tmpReg,0x1);
+            }
+         else
+            {
+            // This is equivalent to doing a `tmpReg & 0x1` (There should be a helper for this to check processor info)
+            TR::InstOpCode::Mnemonic opCode = cg->getS390ProcessorInfo()->supportsArch(TR_S390ProcessorInfo::TR_zEC12) ? TR::InstOpCode::RISBGN : TR::InstOpCode::RISBG;
+            generateRIEInstruction(cg, opCode, node, tmpReg, tmpReg, 63, 191, 0);
+            }
+         //now left shift by x to get to the position u want
+         if (x > 0)
+            generateRSInstruction(cg, TR::InstOpCode::getShiftLeftLogicalSingleOpCode(), node, tmpReg, tmpReg, x);
+
+         // now OR the result into the resultReg
+         generateRRInstruction(cg, TR::InstOpCode::getOrRegOpCode(), node, resultReg, tmpReg);
+         }
+      }
+   else
+      {
+      auto lengthReg = cg->evaluate(length);
+      auto indexReg = cg->allocateRegister(TR_GPR);
+
+      TR::RegisterDependencyConditions* deps = generateRegisterDependencyConditions((uint8_t)0, (uint8_t)4, cg);
+      deps->addPostCondition(addrReg, TR::RealRegister::AssignAny);
+      deps->addPostCondition(indexReg, TR::RealRegister::AssignAny);
+      deps->addPostCondition(valueReg, TR::RealRegister::AssignAny);
+      deps->addPostCondition(tmpReg, TR::RealRegister::AssignAny);
+
+      TR::LabelSymbol *startLabel = generateLabelSymbol(cg);
+      TR::LabelSymbol *endLabel = generateLabelSymbol(cg);
+      startLabel->setStartInternalControlFlow();
+      endLabel->setEndInternalControlFlow();
+
+      // Load array length into index and test to see if it's already zero
+      // cc=0 if register value is 0, cc=1 if value < 1, cc=2 if value>0
+      generateRRInstruction(cg, TR::InstOpCode::LTGR, node, indexReg, lengthReg); 
+      
+      //start of internal control flow 
+      generateS390LabelInstruction(cg,TR::InstOpCode::LABEL, node, startLabel);
+
+      // now subtract 1 from indexReg 
+      generateRILInstruction(cg, node->getDataType() == TR::Int64 ? TR::InstOpCode::AGFI : TR::InstOpCode::AFI, node, indexReg, -1);
```

Comment:
Can we use

```
generateS390ImmOp(cg, TR::InstOpCode::getAddOpCode(), node, indexReg, indexReg, -1);
```

here instead and let the helper pick the best instruction? All add instructions produce the same condition code results.

Reply:
Done.